system/jlib/jlz4.cpp

```diff
@@ -197,6 +198,419 @@ class CLZ4Compressor final : public CFcmpCompressor
     }
 };
 
+//---------------------------------------------------------------------------------------------------------------------
+
+//Enable the following to debug details of the stream compression
+//#define LZ4LOGGING
+
+
+//Helper functions for storing "packed" numbers, where the top bit indicates the number is continued in the next byte
+static unsigned sizePacked(size32_t value)
+{
+    unsigned size = 1;
+    while (value >= 0x80)
+    {
+        value >>= 7;
+        size++;
+    }
+    return size;
+}
+
+static void writePacked(byte * out, size32_t & offset, unsigned __int64  value)
+{
+    constexpr unsigned maxBytes = 9;
+    unsigned index = maxBytes;
+    byte result[maxBytes];
+
+    result[--index] = value & 0x7f;
+    while (value >= 0x80)
+    {
+        value >>= 7;
+        result[--index] = (value & 0x7f) | 0x80;
+    }
+
+    size32_t len = maxBytes - index;
+    memcpy(out+offset, result + index, len);
+    offset += len;
+}
+
+static unsigned readPacked32(const byte * & cur)
+{
+    byte next = *cur++;
+    unsigned value = next;
+    if (unlikely(next >= 0x80))
+    {
+        value = value & 0x7f;
+        do
+        {
+            next = *cur++;
+            value = (value << 7) | (next & 0x7f);
+        } while (next & 0x80);
+    }
+    return value;
+}
+
+
+
+/*
+ * The serialized stream compressed data has the following format:
+ *
+ * 4-bytes - length of the uncompressed data
+ * List of packed lengths of the compressed blocks of data, followed by a zero (=packed int 0)
+ * Blocks of compressed data concatenated together
+ *    packedlength - length of the compressed data
+ *    data         - block of compressed data
+ * 0 byte (i.e. zero packed length)
+ * <fill data> - uncompressed length - length already expanded
+ *
+ * The compressor builds up the compressed data into a pair of output buffers
+ * When recompressing the other buffer is used as the target, so that the previous
+ * compression is preserved in the unusual situation where recompressing makes it larger.
+*/
+class CLZ4StreamCompressor final : public CSimpleInterfaceOf<ICompressor>
+{
+public:
+    CLZ4StreamCompressor(const char * options, bool _hc) : hc(_hc)
+    {
+        auto processOption = [this](const char * option, const char * value)
+        {
+            if (strieq(option, "hclevel"))
+                hcLevel = atoi(value);
```

Comment:
Do we need a bounds check on this to keep it between 1 and LZ4HC_CLEVEL_MAX? Or at least add a comment noting the valid range?

Reply:
I will add that.